crates/bifrost/src/read_stream.rs

```diff
@@ -106,47 +147,263 @@ impl LogReadStream {
+
+            match state {
+                StateProj::New => {
+                    let find_loglet_fut = Box::pin(
+                        bifrost_inner.find_loglet_for_lsn(*this.log_id, *this.read_pointer),
+                    );
+                    // => Find Loglet
+                    this.state.set(State::FindingLoglet { find_loglet_fut });
+                }
+
+                // Finding a loglet and creating the loglet instance through the provider
+                StateProj::FindingLoglet { find_loglet_fut } => {
+                    let loglet = match ready!(find_loglet_fut.poll(cx)) {
+                        Ok(loglet) => loglet,
+                        Err(e) => {
+                            this.state.set(State::Terminated);
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                    };
+                    // create sub-stream to read from this loglet.
+                    let create_stream_fut =
+                        Box::pin(loglet.create_wrapped_read_stream(*this.read_pointer));
+                    // => Create Substream
+                    this.state
+                        .set(State::CreatingSubstream { create_stream_fut });
+                }
+
+                // Creating a new substream
+                StateProj::CreatingSubstream { create_stream_fut } => {
+                    let substream = match ready!(create_stream_fut.poll(cx)) {
+                        Ok(substream) => substream,
+                        Err(e) => {
+                            this.state.set(State::Terminated);
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                    };
+                    let safe_known_tail = substream.tail_lsn();
+                    // If the substream's tail is unknown, we will need to watch the tail updates.
+                    let tail_watch = if safe_known_tail.is_none() {
+                        Some(substream.loglet().watch_tail())
+                    } else {
+                        None
+                    };
+                    // => Start Reading
+                    this.substream.set(Some(substream));
+                    this.state.set(State::Reading {
+                        safe_known_tail,
+                        tail_watch,
+                    });
+                }
+
+                // Reading from the current substream
+                StateProj::Reading {
+                    safe_known_tail,
+                    tail_watch,
+                } => {
+                    // Continue driving the substream
+                    // This depends on whether we know its tail (if sealed), or if we the value of
+                    // `safe_known_tail` higher than the `read_pointer` of the substream.
+                    let Some(substream) = this.substream.as_mut().as_pin_mut() else {
+                        panic!("substream must be set at this point");
+                    };
```

Comment:
Could `substream` become part of `State::Reading`? It seems that we set the `state` and `substream` field always in tandem.

Reply:
It started this way, but then moving them across states became a nightmare (because the variant will self-reference in that case)

Reply:
Yeah, I guess the field in the variant would need to be `Option<...>` so that one can move out of it into the other variant. This is probably not really nice.

Reply:
The problem is more related to when creating a future from the loglet because all loglet functions hold a reference to self (async-trait). The future (like creating the substream) will need a reference to the substream for as long as it's alive.

Reply:
Hmm, when I tried moving the `substream` to the state, I did not encounter this problem. We seem to only call `poll_next` on the `substream` and do not create other futures from it.

Reply:
Explained offline. The historical context was that I was doing find_tail() on the loglet (owned by the substream) which we might need to do for background check when the stream is idle in the future. I'm leaning towards leaving it as is to make this easier since the Option::unwrap is not a lot prettier as discussed.